clang/test/dpct/helper_files_ref/include/util.hpp

```diff
@@ -346,6 +346,50 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
                                      : id);
 }
 
+/// The logical-group is a logical collection of some work-items within a
+/// work-group. The size of a logical-group can be 2, 4, 8, 16 or 32.
+class logical_group {
+  cl::sycl::group<3> _g;
+  uint32_t _logical_group_size;
+  uint32_t _group_linear_range_in_parent;
+public:
+  /// Dividing \p parent_group into several logical-groups.
+  /// \param [in] parent_group The group to be divided.
+  /// \param [in] size The logical-group size.
+  logical_group(cl::sycl::group<3> parent_group, uint32_t size)
+      : _g(parent_group), _logical_group_size(size) {
+    _group_linear_range_in_parent =
```

Comment:
we should do extra checker on the size. (<=sub-group-size)

Reply:
We cannot throw exception in device code.
So I set the logical group size to 1 if it is > sub-group size and documented in the comment.